### PR TITLE
security: OpenVPN sanitizer redaction, pfSense validator atomicity, plugin loader preflight

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Hardens the sanitizer, plugin loader, and parser-injection surface before v1.5 tags. Closes a sanitizer gap that leaked OpenVPN TLS-auth keys, protects the pfSense validator injection point from malicious plugin `init()` stomp, adds a preflight pass in front of `plugin.Open()`, and disclosures the dynamic-plugin trust model honestly to operators.

**Impact:** 35 files changed, +2,190 / -153. Most growth is in new test files (plugin hardening, race detection, preflight scenarios) and new implementation files for the preflight and POSIX/Windows splits.

**Risk:** 🟠 High-to-medium — one breaking public-API change (`pfsense.ValidateFunc` → `pfsense.SetValidator`, pre-v1.5 free window) and several security-sensitive code paths. All changes are additive or behaviorally backward-compatible for non-plugin invocations.

**Review time:** ~45 minutes. Recommended review order:

1. `internal/sanitizer/` — OpenVPN redaction gap (highest-severity fix)
2. `pkg/parser/pfsense/parser.go` + `export_test.go` — validator atomicity (breaking API)
3. `internal/audit/plugin_preflight.go` + POSIX/Windows split files — loader preflight
4. `cmd/shared_flags.go` + `cmd/audit.go` + `docs/user-guide/commands/audit.md` — trust-model disclosure
5. `internal/sanitizer/sanitizer.go` + `pkg/parser/factory.go` + audit callsites — reflect-map warning, cancellation contract, log-redaction gates

### OpenVPN TLS-auth key redaction

`opnDossier sanitize` now redacts OpenVPN `<tls>` and `<StaticKeys>` XML elements. Path-anchored patterns (`openvpn.tls`, `openvpn.statickeys`) identify the real OpenVPN paths without colliding with unrelated `<tls>` tags in Suricata IDS and IPsec charon syslog configurations. A new `IsOpenVPNStaticKey` value detector matches the OpenVPN static-key PEM envelope (`-----BEGIN OpenVPN Static key V1-----`). Previously, operators running sanitize on configs containing OpenVPN TLS-auth material leaked the raw HMAC keys to stdout. Regression coverage: 4 fixtures × 3 sanitize modes plus 2 false-positive cases.

### pfSense validator injection, atomic one-shot

`pkg/parser/pfsense.ValidateFunc` (exported `var`) is **removed** and replaced by `pkg/parser/pfsense.SetValidator`, guarded by `sync.Once` over an `atomic.Pointer`. The public API is now:

```go
// Before (pre-v1.5, removed)
pfsense.ValidateFunc = myValidator

// After
pfsense.SetValidator(myValidator)
```

Second and subsequent `SetValidator` calls are silently dropped. This closes a data race between plugin `init()` functions and `ParseAndValidate` readers under `-race`, and — intentionally — prevents a malicious dynamic plugin from stomping the validator at package-init time. `ResetValidatorForTesting` lives in `export_test.go` so it stays out of the public API; production builds cannot reach it. Matches the injection discipline the OPNsense parser already has via `XMLDecoder.ParseAndValidate`.

**This is a breaking public-API change under the pre-v1.5 free window.** Downstream consumers must migrate `ValidateFunc = fn` → `SetValidator(fn)`.

### Plugin loader preflight + structured audit log

Before `plugin.Open()` runs on any candidate `.so`, a new `runPluginPreflight` function enforces:

1. **Absolute path required** (cross-platform). `LoadDynamicPlugins` normalizes `--plugin-dir` via `filepath.Abs` at function entry, so `--plugin-dir ./plugins` works correctly and the audit log records the canonical path.
2. **Symlinks rejected outright** (cross-platform via `os.Lstat`). `plugin.Open` follows symlinks, which would bypass the remaining checks against the link target.
3. **Group/world-writable plugin files rejected** (POSIX only, `FileMode.Perm() & 0o022`).
4. **Group/world-writable plugin directories rejected** (POSIX only).
5. **SHA-256 digest computed** with a 64 MiB read cap (`hashFileSizeCapped`). Files exceeding the cap are rejected rather than silently truncated.

Every load attempt emits a structured audit log entry — INFO for accepted, WARN for rejected — carrying `plugin`, `path`, `sha256`, `mode`, `owner_uid`, `mtime`, `size_bytes`, `verdict`, and `reason`. Operators gain a forensic trail even when `plugin.Open` subsequently fails.

POSIX and Windows implementations are split via build tags:

- `internal/audit/plugin_preflight_posix.go` (`//go:build !windows`) — full permission-bit enforcement via `syscall.Stat_t`, owner UID extraction via `Stat_t.Uid`
- `internal/audit/plugin_preflight_windows.go` (`//go:build windows`) — permission bits are skipped (NTFS permissions do not map onto POSIX `FileMode.Perm()` bits); symlink and absolute-path checks still run; `owner_uid=unavailable` in the audit log

### `--plugin-dir` trust-model disclosure

`--plugin-dir` help text is expanded and now reads:

> Directory containing dynamic .so compliance plugins. Plugins run with full process privileges; signatures are not verified. Do not point at untrusted-writable directories. See GOTCHAS §2.5 and docs/user-guide/commands/audit.md § Dynamic Plugin Security.

At invocation time, `audit PreRunE` emits a matching stderr warning whenever `--plugin-dir` is non-empty, via a shared `warnPluginDirTrustModel(io.Writer, string)` helper so every command that accepts `--plugin-dir` stays aligned. README and the user-guide audit command doc now carry a "Dynamic Plugin Security" subsection that documents the threat model (full process privileges, no signature verification, opt-in only) alongside the preflight checks listed above.

### Reflect-map redaction warning

`internal/sanitizer/sanitizer.go` adds `Sanitizer.SetLogger` plus a reflection-path warning emitted from `SanitizeStruct` whenever the walker encounters a `map[K]struct{...}` or `map[K]*struct{...}` value. Go disallows in-place mutation of map-struct values, so the walker has always skipped them silently. The warning surfaces the gap so a future schema routing secrets through a struct-valued map is detected at runtime rather than shipped as cleartext. The raw-XML `SanitizeXML` path is unaffected. Full structural redaction of struct-valued maps will land separately under the tag-based-redaction refactor.

### Goroutine-leak contract on `peekRootElementBounded`

`pkg/parser/factory.go`'s `peekRootElementBounded` now documents its cancellation contract in-file: the function returns `ctx.Err()` immediately on cancellation, but its internal goroutine only exits when the next read unblocks. Library consumers supplying readers that can block indefinitely (sockets, fifos, long-polling HTTP bodies) MUST cancel the context to release the goroutine. CLI consumers wrapping `*os.File` see prompt `io.EOF` and are unaffected. No safety-net watchdog was added — `peekRootElementBounded` does not own the reader it receives, and closing a caller-owned reader would corrupt caller state. A 100 ms regression test against a hung `io.Pipe` exercises the contract.

### Log redaction gate

`debug.Stack()` dumps at `internal/audit/plugin.go`, `cmd/audit.go`, and `internal/processor/processor.go` are now gated behind `logger.IsVerbose()`. Go stack traces include package paths, so unredacted stacks can leak customer-specific plugin identifiers (e.g., `acmecorp-pci-plugin.RunChecks`) into centralized logs and expose compliance posture. Default log mode emits a structured error without the stack; verbose mode retains the full stack for debugging. Nil-safe `IsVerbose()` handles the panic-recovery path edge cases.

## Test Plan

- [x] `just ci-check` passes (lint + tests + race detector + integration)
- [x] `just test-race ./pkg/parser/pfsense/...` — validator atomicity pinned via `atomic.Pointer`; new concurrent-writer race test
- [x] `just test-race ./internal/audit/...` — plugin loader preflight exercises concurrent load
- [x] Cross-platform compile: `GOOS=linux go build ./internal/audit/...` and `GOOS=windows go build ./internal/audit/...` both succeed
- [x] `pre-commit run --all-files` passes
- [x] API snapshot goldens regenerated for `pkg/parser/pfsense` — `SetValidator` present, `ValidateFunc` removed
- [x] Plugin loader test coverage: implicit-missing-dir silence, explicit-missing-dir error, non-directory path, non-.so entries skipped, size-cap enforcement, symlink rejection, permission-bit rejection, relative-path normalization, SHA-256 digest of files exceeding cap
- [x] Sanitizer test coverage: OpenVPN 4 fixtures × 3 modes, 2 false-positive cases (Suricata, IPsec charon), `IsOpenVPNStaticKey` PEM-envelope matcher

## Out of Scope

- **Further plugin loader hardening** — owner-UID verification against an allowlist, operator-configurable size cap, path denylist, filename allowlist, optional SHA-256 manifest, seccomp/landlock sandboxing. Tracked in GOTCHAS §2.5 for post-v1.5.
- **Tag-based redaction architecture** — structural replacement for the map-struct gap that the new warning surfaces. A 2–3 day refactor that warrants its own focused effort.

### CODE REVIEW

File: internal/audit/plugin_preflight.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+// runPluginPreflight performs the Phase A hardening checks documented in
+// GOTCHAS §2.5. It returns a pluginPreflightResult describing the outcome; on
+// success the caller should proceed with plugin.Open, on rejection the
+// caller should record the embedded error as a PluginLoadError.
+//
+// The checks, in order:
+//  1. The path must be absolute (cross-platform).
+//  2. Lstat the path. A symlink is rejected outright because plugin.Open
+//     follows links (POSIX only — Windows symlink semantics differ and are
+//     skipped at runtime).
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_preflight.go]: > The preflight check-list comment says symlink handling is "POSIX only … skipped at runtime", but the implementation below explicitly rejects symlinks on Windows too.

Addressed in a084f9e: the symlink check is cross-platform (via `os.Lstat`); the POSIX-only caveat only applies to the mode-bit checks (group/world-writable). Updated the comment to say "A symlink is rejected outright on every platform because plugin.Open follows links."

File: pkg/parser/factory.go [hunk 2]
```diff
@@ -193,13 +193,35 @@ type peekResult struct {
...
+// reuse the reader after cancellation, and not every reader is an
+// [io.Closer]). The cancellation contract is the single mechanism callers
+// use to bound goroutine lifetime.
+//
+// CLI callers wrap *os.File readers which return io.EOF promptly on EOF, so
+// the goroutine exits naturally in that path. Library consumers supplying
+// readers that can block indefinitely (sockets, fifos, long-polling HTTP
+// bodies) MUST cancel the context to release the goroutine. See AGENTS.md
+// §1.1 for why we never use t.Parallel() in tests that exercise this path
+// against package-level state.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/factory.go]: > The cancellation-contract comment references "AGENTS.md §1.1", but `AGENTS.md` doesn't have a §1.1 section; the `t.Parallel()` / global-state guidance lives in `GOTCHAS.md` (### 1.1).

Addressed in a084f9e: corrected the cross-reference to `GOTCHAS.md §1.1`.

File: docs/user-guide/commands/audit.md [hunk 3]
```diff
@@ -81,6 +81,31 @@ opndossier audit config.xml --mode blue --plugin-dir /opt/plugins
...
+
+Before enabling `--plugin-dir`:
+
+- Restrict filesystem permissions on the plugin directory so only trusted operators can drop files in it. Avoid world-writable paths, `/tmp`, and shared CI scratch directories.
+- Review plugin source code and build plugins from known-good checkouts.
+- Keep the plugin directory out of paths that untrusted processes (user shells on a shared host, CI runners accepting external PRs) can write to.
+- Pin plugin builds to the same Go toolchain and module set as the opnDossier binary — a version mismatch is caught at load time, but only after the `.so` has been mapped into the process.
+
+Planned hardening (e.g. stricter symlink handling and load-time provenance checks) is tracked as Phase A of the plugin sandboxing work. Until then, the guidance above is the full mitigation.
+
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply docs/user-guide/commands/audit.md]: > This section still states that "Symlinks are followed" and that Phase A "stricter symlink handling" is only planned, but the loader now runs a preflight that rejects symlink plugins before `plugin.Open`.

Addressed in a084f9e: rewrote the Dynamic Plugin Security section to document the Phase A preflight accurately — symlink rejection, group/world-writable file and directory checks (POSIX only), absolute-path normalization, SHA-256 audit trail, and platform caveats. Also dropped the stale "default ./plugins directory" claim (loading is opt-in, requires explicit `--plugin-dir`). Phase B hardening items are listed as tracked for post-v1.5.

File: pkg/parser/pfsense/parser_test.go [hunk 4]
```diff
@@ -1558,6 +1569,109 @@ func TestParser_ParseAndValidateWithValidator(t *testing.T) {
...
+
+	for range numWriters {
+		go func() {
+			defer installed.Done()
+			<-writerBarrier
+			pfsense.SetValidator(func(_ *pfsenseSchema.Document) error {
+				return nil
+			})
+		}()
+	}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/pfsense/parser_test.go:1645]: > `for range numWriters` does not compile (you can only range over arrays/slices/maps/channels/strings).

Not addressing: `for range N` over integers has been valid Go syntax since 1.22 (see [Go 1.22 release notes — "Range over integers"](https://go.dev/doc/go1.22#language)). This project's `go.mod` pins `go 1.26`, so this loop form compiles cleanly. CI and local `go test` both pass with these loops as-written.

File: pkg/parser/pfsense/parser_test.go [hunk 5]
```diff
@@ -1558,6 +1569,109 @@ func TestParser_ParseAndValidateWithValidator(t *testing.T) {
...
+
+	for range numReaders {
+		go func() {
+			defer readerWG.Done()
+			<-readerBarrier
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/pfsense/parser_test.go:1650]: > `for range numReaders` does not compile for the same reason as the writers loop.

Not addressing: same rationale as the writers loop one line above — `for range N` over integers has been valid Go syntax since 1.22, and this project is on `go 1.26`. The test compiles and passes.

File: internal/audit/plugin.go [hunk 6]
```diff
@@ -169,6 +169,20 @@ func (pr *PluginRegistry) LoadDynamicPlugins(
...
+		// is ever called. A structured audit-log entry is emitted for every
+		// attempt, accepted or rejected.
+		preflight := runPluginPreflight(path)
+		logPreflight(ctxLogger, preflight)
+
+		if preflight.verdict != pluginVerdictAccepted {
+			failures = append(failures, PluginLoadError{Name: entry.Name(), Err: preflight.err})
+
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin.go:232]: > `runPluginPreflight` rejects non-absolute paths, but `LoadDynamicPlugins` passes `filepath.Join(dir, entry.Name())` without normalizing `dir`. This means a common invocation like `--plugin-dir ./plugins` will cause every plugin to be rejected as a "relative path".

Addressed in a084f9e: `LoadDynamicPlugins` now resolves `dir` to an absolute path via `filepath.Abs` at the top of the function, before `os.ReadDir` and before any preflight runs. The audit log records the normalized absolute path. The relative-path preflight check remains as defense-in-depth for any future caller that bypasses `LoadDynamicPlugins`. The integration test was renamed `TestLoadDynamicPlugins_NormalizesRelativeDir` and now asserts the new behavior (preflight accepts, loader fails instead of preflight rejection). The direct-preflight unit test for relative-path rejection at line 436 of `plugin_hardening_test.go` is unchanged.

File: internal/audit/plugin_preflight.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	"time"
+
+	"github.com/EvilBit-Labs/opnDossier/internal/logging"
+)
+
+// Phase A plugin loader hardening constants. These bound the preflight
+// behaviour and are exported as package-level const so tests and docs can
+// reference a single source of truth. Phase B follow-ups (owner-UID check,
+// path denylist, filename allowlist, SHA-256 manifest) are tracked for
+// post-v1.5; see GOTCHAS §2.5.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_preflight.go]: > The comment says these Phase A constants are "exported", but they are unexported identifiers (`pluginWritableMask`, `pluginMaxSize`, etc.).

Addressed in a084f9e: reworded to "These package-level constants bound the preflight behaviour so code, tests in this package, and documentation comments can reference a single source of truth." Keeping them unexported is intentional — they're implementation-detail bounds for Phase A, not a public API.

File: CHANGELOG.md [hunk 8]
```diff
@@ -125,6 +137,30 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+### Security
+
+- Sanitizer now redacts OpenVPN `<tls>` and `<StaticKeys>` XML elements
+  and detects the OpenVPN static-key PEM envelope. Previously, operators
+  running `opnDossier sanitize` on configs containing OpenVPN TLS-auth
+  material leaked the raw HMAC keys (SEC-H1 from comprehensive review).
+- pfSense validator protected against stomp by malicious dynamic plugin
+  `init()` code. `pfsense.ValidateFunc` renamed to unexported
+  `validateFunc`; new `pfsense.SetValidator` guarded by `sync.Once`.
+  Matches OPNsense's `XMLDecoder.ParseAndValidate` DI pattern. Fixes
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply CHANGELOG.md]: > The changelog names the wrong replacement symbol.

Addressed in a794af7: CHANGELOG entry now says the public `pfsense.ValidateFunc` var was **removed** and replaced by `pfsense.SetValidator` (guarded by `sync.Once` over an `atomic.Pointer` holder), rather than "renamed to unexported validateFunc".

File: docs/security/security-assurance.md [hunk 9]
```diff
@@ -24,24 +24,28 @@ opnDossier is an OPNsense configuration parser, auditor, and reporting tool. Its
...
+| ----- | -------------------------------------------------------------------------- | ---------------- |
+| AV-1  | XXE or entity expansion in crafted config.xml                              | SR-2, SR-6       |
+| AV-2  | Deeply nested XML elements cause stack overflow or resource exhaustion     | SR-1, SR-6       |
+| AV-3  | Extremely large config.xml causes memory exhaustion                        | SR-6             |
+| AV-4  | CLI argument with path traversal writes reports to unintended locations    | SR-3             |
+| AV-5  | Crafted XML triggers panic in parser or schema mapping                     | SR-1             |
+| AV-6  | Error messages include raw credential values from config                   | SR-5             |
+| AV-7  | Compromised Go module introduces malicious code                            | SR-4             |
+| AV-8  | Crafted IPsec XML elements exploit pfSense IPsec parser paths              | SR-1, SR-2, SR-6 |
+| AV-9  | Malicious plugin init() stomps validator before preflight                  | SR-4             |
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply docs/security/security-assurance.md]: > AV-9 describes an impossible ordering.

Addressed in a794af7: AV-9 now reads "Malicious plugin init() stomps validator during plugin.Open() load" — init() fires when plugin.Open loads the .so, which is *after* `runPluginPreflight` has already run per the LoadDynamicPlugins flow.

File: internal/audit/plugin_hardening_test.go [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+		t.Fatal("LoadDynamicPlugins() expected error for symlink plugin")
+	}
+
+	if result.Failed() != 1 {
+		t.Fatalf("Failed() = %d, want 1", result.Failed())
+	}
+
+	if !strings.Contains(result.Failures[0].Error(), "symlink") {
+		t.Errorf("expected 'symlink' in failure error, got: %v", result.Failures[0].Err)
+	}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_hardening_test.go:110]: > Inconsistent error reference in assertion vs log.

Addressed in a794af7: both the `strings.Contains` check and the `t.Errorf` formatting now use `result.Failures[0].Error()` consistently.

File: internal/audit/plugin_hardening_test.go [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+	t.Cleanup(func() {
+		// Restore permissions so t.TempDir cleanup can remove the dir on
+		// platforms where the test user cannot unlink files inside a
+		// world-writable directory without write on the parent.
+		//nolint:gosec // 0o700 is the default t.TempDir mode; explicit restore.
+		if err := os.Chmod(pluginDir, 0o700); err != nil {
+			t.Logf("warning: failed to restore plugin dir permissions: %v", err)
+		}
+	})
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_hardening_test.go]: > Nolint directive placement should be on separate line above call.

Addressed in a794af7: moved the inline `//nolint:gosec` on the `os.Chmod(pluginDir, 0o777)` call to a separate line above. Also added a `//nolint:gosec` on the cleanup `os.Chmod(pluginDir, 0o700)` — gosec G302 flags any chmod literal > 0o600 regardless of whether the mode is actually restrictive.

File: internal/audit/plugin_preflight.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	// Compute SHA-256 before handing the file to plugin.Open. Do this last
+	// so cheap rejections short-circuit without reading file contents. The
+	// read is capped at pluginMaxSize to bound memory and I/O.
+	digest, hashErr := hashFileSizeCapped(path, pluginMaxSize)
+	if hashErr != nil {
+		res.reason = "sha256 computation failed"
+		res.err = fmt.Errorf("failed to hash plugin %q: %w", path, hashErr)
+		return res
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_preflight.go:177]: > Reject non-regular plugin entries before hashing them.

Addressed in a794af7: `runPluginPreflight` now rejects non-regular files via `info.Mode().IsRegular()` immediately after the symlink check (cross-platform). FIFOs, sockets, and device nodes named `*.so` no longer reach `hashFileSizeCapped`. Added a `"non-regular file is rejected"` subtest to `TestRunPluginPreflight_Unit` using a directory named `*.so` (portable across platforms, no `syscall.Mkfifo` dependency). Also updated the preflight doc-list to reflect the new six-step sequence and added a new bullet to GOTCHAS.md §2.5 Phase A.

File: internal/sanitizer/sanitizer_reflect_test.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	if err := s.SanitizeStruct(container); err != nil {
+		t.Fatalf("SanitizeStruct() error = %v", err)
+	}
+
+	// String map values are still redacted in place.
+	if got := container.Passwords["password"]; got == "plaintext-secret" {
+		t.Errorf("password not redacted for string-valued map; got %q", got)
+	}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/sanitizer/sanitizer_reflect_test.go]: > Assert the exact redacted value here.

Addressed in a794af7: the test now asserts `container.Passwords["password"]` equals the exact placeholder `"[REDACTED-PASSWORD]"` emitted by the password rule in `internal/sanitizer/rules.go`, rather than just checking inequality against the plaintext.

File: internal/sanitizer/sanitizer.go [hunk 14]
```diff
@@ -363,6 +376,24 @@ func (s *Sanitizer) sanitizeReflect(v reflect.Value, path string) error {
...
+		if elemKind == reflect.Struct || elemKind == reflect.Ptr {
+			if s.logger != nil {
+				s.logger.Warn(
+					"sanitize reflect: skipping map with struct/pointer values",
+					"path", path,
+					"type", v.Type().String(),
+				)
+			}
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/sanitizer/sanitizer.go:396]: > The struct/pointer-valued map guard only checks the *declared* map element kind ... For maps with interface element types ... struct/pointer values are still silently skipped

Addressed in a794af7: after the declared-kind check, `SanitizeStruct` now inspects interface-typed maps (`reflect.Interface` element kind). It unwraps interface values via `Elem()` and, if the concrete dynamic kind is `Struct` or `Ptr`, emits the same skip warning naming both the declared map type and the concrete type, then returns. The check breaks after the first match — one warning per offending map is enough signal for schema triage.

File: pkg/parser/factory.go [hunk 15]
```diff
@@ -193,13 +193,35 @@ type peekResult struct {
...
+// consumed bytes. It returns the root element name, a reader that replays the
+// buffered bytes followed by the remainder of r, and any error.
+//
+// CANCELLATION CONTRACT: Callers must ensure the supplied ctx is eventually
+// cancelled. On ctx.Done(), this function returns ctx.Err() immediately, but
+// the inner goroutine itself only exits when the next token read unblocks —
+// the ctx-wrapped reader (see [newCtxReader]) surfaces ctx.Err() on its next
+// Read call, which is how the goroutine is unstuck. If the supplied reader
+// never yields (e.g., a hung network stream) AND the ctx is never cancelled,
+// the goroutine leaks and retains up to DefaultMaxInputSize bytes in its
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/factory.go]: > The cancellation-contract comment implies the ctx-wrapped reader will "unstick" the goroutine by surfacing ctx.Err() on its next Read call. That's only true once the underlying Read returns

Addressed in a794af7: tightened the wording to match the actual behavior. The comment now reads: "The ctx-wrapped reader (see [newCtxReader]) returns ctx.Err() on a *subsequent* Read call after the underlying blocked Read has returned; it does not interrupt an already-blocked Read."

File: GOTCHAS.md [hunk 16]
```diff
@@ -55,6 +55,20 @@ Reclassified info-severity controls (e.g., FIREWALL-003 "Message of the Day") pa
...
+
+- **Symlinks rejected** via `os.Lstat` + `os.ModeSymlink` check. `plugin.Open` follows links, so an attacker with write access to the plugin directory could otherwise point a `.so` at anything on the filesystem.
+- **Group/world-writable plugin files rejected** when `info.Mode().Perm()&0o022 != 0` (POSIX only). Closes CWE-732 where another local account could swap the file between audits.
+- **Group/world-writable container directory rejected** via a second `os.Stat` on `filepath.Dir(path)` (POSIX only). The file bits alone are not enough — a writable parent lets an attacker unlink and replace the plugin.
+- **Absolute paths required** via `filepath.IsAbs` (cross-platform). Operators must pass a fully-qualified `--plugin-dir`; relative paths are rejected so audit logs unambiguously identify the artifact.
+- **Structured audit log per load attempt**: INFO for accepted loads, WARN for rejections, with fields `plugin`, `path`, `sha256`, `mode`, `owner_uid`, `mtime`, `size_bytes`, `verdict`, `reason`. The SHA-256 is computed with a 64 MiB read cap so a pathological `.so` cannot exhaust memory during preflight.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply GOTCHAS.md]: > The Phase A bullet "Absolute paths required … Operators must pass a fully-qualified --plugin-dir; relative paths are rejected" is now inaccurate

Addressed in a794af7: reworded GOTCHAS.md §2.5 Phase A "Absolute paths" bullet. It now says the absolute-path check at the preflight boundary is defense-in-depth against callers that bypass LoadDynamicPlugins, while relative `--plugin-dir` operator inputs are accepted and normalized via `filepath.Abs` before preflight. Also updated the structured-audit-log bullet to reflect that the logged `path` is the normalized absolute artifact path, and fixed the SHA-256-cap rationale (bounds I/O/CPU, not memory — the hasher streams).

File: pkg/parser/pfsense/parser_test.go [hunk 17]
```diff
@@ -1558,6 +1569,109 @@ func TestParser_ParseAndValidateWithValidator(t *testing.T) {
...
+	const (
+		numWriters = 16
+		numReaders = 16
+	)
+
+	// Every writer installs the SAME validator (identical function value is
+	// not required by the spec — only "some installed function wins and
+	// stays installed"). Because we cannot reliably compare func values,
+	// each writer just increments the shared counter; the race test
+	// succeeds if no race is flagged and the reader path executes cleanly.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/pfsense/parser_test.go]: > This comment says "each writer just increments the shared counter", but there is no counter in the test anymore.

Addressed in a794af7: removed the stale "shared counter" sentence. Comment now describes the actual assertion — no race flagged by `-race` when writers call SetValidator concurrently while the reader path (ParseAndValidate) executes.

File: internal/audit/plugin_hardening_test.go [hunk 18]
```diff
@@ -0,0 +1,711 @@
...
+// can assert on the structured result without going through the full
+// LoadDynamicPlugins pipeline. Guards basic invariants: accepted paths carry
+// a sha256 digest, rejections carry a non-nil err, and the verdict string
+// stays within the two documented values.
+func TestRunPluginPreflight_Unit(t *testing.T) {
+	t.Parallel()
+
+	if runtime.GOOS == goosWindows {
+		t.Skip("POSIX permission semantics required")
+	}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_hardening_test.go]: > Don't skip the platform-neutral preflight cases on Windows.

Addressed in a794af7: the top-level `runtime.GOOS == goosWindows` skip was removed from `TestRunPluginPreflight_Unit`. The relative-path, missing-file, and non-regular-file subtests now run cross-platform. Only the accepted-plugin subtest retains the Windows skip (it depends on POSIX permission semantics for a deterministic accepted path). The POSIX-gated permission-bit tests live in separate standalone functions that keep their own Windows skips.

File: internal/audit/plugin_hardening_test.go [hunk 19]
```diff
@@ -0,0 +1,711 @@
...
+		t.Fatalf("expected 1 PluginInfo entry, got %d", len(result.PluginInfo))
+	}
+	info, ok := result.PluginInfo["panicking-plugin-verbose"]
+	if !ok {
+		t.Fatal("expected PluginInfo entry for panicked plugin")
+	}
+	if !strings.Contains(info.Version, "panicked") {
+		t.Errorf("expected version to contain 'panicked', got %q", info.Version)
+	}
+}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_hardening_test.go:701]: > This test doesn't actually verify the verbose logging branch.

Addressed in a794af7: `TestRunComplianceChecks_PanicVerboseLogging` now uses `bufferLogger()` to capture log output and asserts the captured buffer contains (1) the "plugin panicked during RunChecks" recovery message, (2) a `stack` field, and (3) the `goroutine ` marker that every `debug.Stack()` payload carries. The `IsVerbose()` gate is now pinned by real output assertions, not just the panic-isolation side effect.

File: internal/audit/plugin_preflight.go [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+	if err != nil {
+		res.reason = "lstat failed"
+		res.err = fmt.Errorf("failed to stat plugin %q: %w", path, err)
+		return res
+	}
+
+	res.mode = info.Mode()
+	res.mtime = info.ModTime()
+	res.sizeBytes = info.Size()
+	res.ownerUID = extractOwnerUID(info)
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_preflight.go:110]: > runPluginPreflight/hashFileSizeCapped will os.Open()+read any non-symlink path ending in .so. On POSIX, a FIFO/socket/device node named "*.so" can make os.Open or io.CopyN block indefinitely

Addressed in a794af7 alongside the sibling thread on this file: added an `info.Mode().IsRegular()` guard in `runPluginPreflight` right after the symlink check. FIFOs, sockets, and device nodes named `*.so` are now rejected before `hashFileSizeCapped` can be reached. Cross-platform subtest covers it.

File: pkg/parser/factory.go [hunk 21]
```diff
@@ -193,13 +193,35 @@ type peekResult struct {
...
+// reuse the reader after cancellation, and not every reader is an
+// [io.Closer]). The cancellation contract is the single mechanism callers
+// use to bound goroutine lifetime.
+//
+// CLI callers wrap *os.File readers which return io.EOF promptly on EOF, so
+// the goroutine exits naturally in that path. Library consumers supplying
+// readers that can block indefinitely (sockets, fifos, long-polling HTTP
+// bodies) MUST cancel the context to release the goroutine. See GOTCHAS.md
+// §1.1 for why we never use t.Parallel() in tests that exercise this path
+// against package-level state.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/factory.go]: > The cancellation-contract comment cites GOTCHAS.md §1.1 as justification for avoiding t.Parallel, but GOTCHAS §1.1 is specifically about cmd/ global Cobra flag variables

Addressed in a794af7: dropped the misleading GOTCHAS.md §1.1 cross-reference. The parser-factory code does not share the cmd/-package global-flag issue that §1.1 documents, so pointing readers there would send them to unrelated guidance.

File: internal/audit/plugin_preflight.go [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+	// (0o020 | 0o002 = 0o022). Any file or directory carrying one of these
+	// bits is writable by an account other than the plugin owner and is
+	// therefore rejected by the preflight.
+	pluginWritableMask os.FileMode = 0o022
+
+	// pluginMaxSize caps the number of bytes we will read while computing
+	// the SHA-256 digest for the audit log. A legitimate compliance plugin
+	// is well under a megabyte; 64 MiB leaves generous headroom while
+	// preventing pathological files from exhausting memory during preflight.
+	// This also partially addresses the Phase B "size cap" objective.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply internal/audit/plugin_preflight.go]: > The pluginMaxSize comment says the cap prevents "pathological files from exhausting memory" during preflight, but hashFileSizeCapped streams into a sha256 hasher and does not buffer file contents in memory.

Addressed in a794af7: reworded the `pluginMaxSize` comment to say the cap "bounds preflight I/O and CPU time and avoids handing very large files to plugin.Open" and explicitly notes the hasher streams rather than buffers (so this is a time/throughput cap, not a memory-allocation cap). Same correction applied to the GOTCHAS.md §2.5 Phase A structured-audit-log bullet.

File: pkg/parser/pfsense/parser_test.go [hunk 23]
```diff
@@ -1515,27 +1516,36 @@ func TestConverter_CronEmptyCommand(t *testing.T) {
...
 
+// minimalValidPfSenseXML is a minimal well-formed pfSense config XML used by
+// the SetValidator regression tests below.
+const minimalValidPfSenseXML = `<?xml version="1.0"?><pfsense><version>19.1</version><system><hostname>test</hostname></system></pfsense>`
+
+// TestParser_ParseAndValidateWithValidator exercises the three observable
+// behaviors of the validator injection: installed-and-succeeds,
+// installed-and-fails, and no-validator fallback. Each subtest resets the
+// sync.Once guard so it can install its own validator; the subtests cannot
+// run in parallel for this reason.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply pkg/parser/pfsense/parser_test.go:1532]: > These SetValidator tests mutate a process-wide validator ... many other tests in this file run with t.Parallel (including at least one that calls ParseAndValidate)

Addressed in a794af7: removed `t.Parallel()` from `TestParser_ParseAndValidate` (the only other ParseAndValidate caller in this file that had it) and added a doc comment explaining why — so it no longer races against `TestParser_ParseAndValidateWithValidator` or `TestParser_SetValidator_StompProtection`, both of which mutate the process-wide validator holder. The SetValidator tests themselves already run serially via `t.Cleanup(pfsense.ResetValidatorForTesting)`.
